### PR TITLE
Address sha1 security compliance

### LINKED ISSUE #10106: [FR] Extend previous work on FIPS compliance
### Willingness to contribute

Yes. I can contribute this feature independently.

### Proposal Summary

From python 3.9 hashlib introduced the usedforsecurity argument:
> Changed in version 3.9: All hashlib constructors take a keyword-only argument usedforsecurity with default value True. A false value allows the use of insecure and blocked hashing algorithms in restricted environments. False indicates that the hashing algorithm is not used in a security context, e.g. as a non-cryptographic one-way compression function.

MLFLOW use hashing in many cases where the purpose is indeed not for security purposes. This should be specifed in the code.

Why you ask?

MLFLOW use MD5 and SHA1 from hashlib, which is not a secure algorithm, but are not specifying that it is for other purposes than security. This is causing issues for organisations following certain security standard. FIPS compliance could be an example.

**My contribution:**
I will attach a PR implementing the similar solution from #9961. I would argue that specifying the kwarg directly could be a cleaner solution, but this is simply a code preference. Another solution can be seen in [this streamlit pull request.](https://github.com/streamlit/streamlit/pull/7527/files)

**Earlier contributions**
I can see that a similar problem was solved in #9961 and addressed in #5959, #9905, #9961.

### Motivation

> #### What is the use case for this feature?

Security compliance

> #### Why is this use case valuable to support for MLflow users in general?
My guess is that pretty much all federal systems in the world would have this issue: "_Federal Information Processing Standards (FIPS) 140-2 is a mandatory standard for the protection of sensitive or valuable data within Federal systems._" - https://www.wolfssl.com/license/fips/

> #### Why is this use case valuable to support for your project(s) or organization?

^^^

> #### Why is it currently difficult to achieve this use case?

As specified in the hashlib documentation both MD5 and SHA1 are proven to have hash collisions: "_Warning Some algorithms have known hash collision weaknesses (including MD5 and SHA1). Refer to [Attacks on cryptographic hash algorithms](https://en.wikipedia.org/wiki/Cryptographic_hash_function#Attacks_on_cryptographic_hash_algorithms) and the [hashlib-seealso](https://docs.python.org/3/library/hashlib.html#hashlib-seealso) section at the end of this document._"

### Details

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/10121?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10121/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10121
```

</p>
</details>

### Related Issues/PRs

Close #10106

### What changes are proposed in this pull request?

`sha1` is considered as insecure and requires `usedforsecurity = false` keyword args when we purposely use it for non security use case. Added [S324](https://databricks.atlassian.net/wiki/spaces/UN/pages/8749065/Glossary+Jargon) to ruff check so we won't accidentally use it directly.

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by harupy — APPROVED** [review]
LGTM!

File: pyproject.toml [hunk 1]
```diff
@@ -49,6 +49,7 @@ select = [
   "UP031",
   "UP032",
   "UP034",
+  "S324", # avoid using insecure hash w/o acknowledgement
```
Comment (harupy) [inline pyproject.toml:52]: Can we move this to line 42?

Comment (harupy) [inline pyproject.toml:52]: I'd remove the comment because we don't have comments on other rules.

Comment (B-Step62) [reply pyproject.toml:52]: makes sense

File: mlflow/pyfunc/__init__.py [hunk 2]
```diff
@@ -1583,7 +1583,7 @@ def batch_predict_fn(pdf, params=None):
                 model_path = os.path.join(
                     tempfile.gettempdir(),
                     "mlflow",
-                    hashlib.sha1(model_uri.encode()).hexdigest(),
+                    sha1(model_uri.encode()).hexdigest(),
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:1586]: Not a blocker, but can we use `insecure_hash.sha1` to emphasize we're using our own module?

Comment (B-Step62) [reply mlflow/pyfunc/__init__.py:1586]: Sure, that sounds cleaner

**Pull request merged**